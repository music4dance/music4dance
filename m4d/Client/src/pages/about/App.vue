<template>
  <page id="app" title="music4dance background info" :breadcrumbs="breadcrumbs">
    <h2>About music4dance</h2>
    <p>
      <b>music4dance</b> is a project that has been floating around in my head
      in various forms since I first started ballroom dancing - so what
      <i>is</i> music4dance, really?
    </p>
    <ul>
      <li>
        <b>music4dance</b> is meant to help find music to dance to as well has
        help figure out what dances can be danced to a piece of music
      </li>
      <li>
        <b>music4dance</b> intended to help DJs find new music to add to their
        playlists as well as for students and teachers to build playlists to
        practice to
      </li>
      <li>
        <b>music4dance</b> meant to help dancers understand the relationship
        between dance and the music that they dance to
      </li>
      <li>
        <b>music4dance</b> a place for people to find music to dance to for
        special occassions like weddings and holidy parties
      </li>
    </ul>
    <p>
      Even to make a dent in any one of those goals is pretty ambitious. That's
      about all I've done is to make a dent. I certainly haven't solved any of
      these problems to my satisfaction. So most of all music4dance is a work in
      progress and a playground for me to experiment with the intersection of
      three of my lifelong passions - music, dance and programming.
    </p>
    <p>
      And while I'm doing that, I hope that music4dance provides at least a
      partial solution to some of the above for you. I'm sure that's true for
      some and I can say that some of the most interesting features of this site
      have started as questions from the community. So please keep using the
      site and <a href="https://music4dance.blog/feedback/">let me know</a> what
      you think.
    </p>
    <p class="neutral">
      Please read my <a href="https://www.music4dance.blog">blog</a> for more
      about the history and motivation behind
      <a href="https://www.music4dance.net">music4dance</a>.
    </p>
    <h2>About me</h2>
    <p>
      I’m an amateur dancer and with some training in music who also happens to
      be a professional software engineer. I love dancing, and I particularly
      love dancing when I enjoy the music that I’m dancing to.
    </p>
    <p>
      My dance background is primarily in Ballroom. I’ve been an amateur
      competitor in the
      <a href="https://www.music4dance.net/dances/american-rhythm"
        >American Rhythm style</a
      >). Recently I've moved more towards social partner dancing. As I was
      starting to work on music4dance in earnest, I dove into tap.
    </p>
    <p>
      Most of my programming time outside of my main day jobs has been related
      to music.
    </p>
    <ul>
      <li>
        When the Macintosh came out, my first thought was - Music Editor. And I
        wrote one - although I never got it to the point where I could sell it.
      </li>
      <li>
        My work study project in college was writing music theory software for
        the Macintosh
      </li>
      <li>
        When I left Corporate America to try my hand at the startup world I
        joined a company that was digitizing sheet music and did the initial
        work to take that online.
      </li>
      <li>
        When I left Corporate America for the second time to rediscover my love
        of programming (and generally find myself) I founded music4dance
      </li>
    </ul>
    <h2>My Resume</h2>
    <p>
      <a href="/Content/DavidGray-Resume.pdf">Printable Version</a>
    </p>
    <p>
      <a href="/Home/Resume">Web Version</a>
    </p>
  </page>
</template>

<script lang="ts">
import Page from "@/components/Page.vue";
import { BreadCrumbItem, infoTrail } from "@/model/BreadCrumbItem";
import { Component, Vue } from "vue-property-decorator";

@Component({
  components: {
    Page,
  },
})
export default class App extends Vue {
  private breadcrumbs: BreadCrumbItem[] = [
    ...infoTrail,
    { text: "About", active: true },
  ];
}
</script>
