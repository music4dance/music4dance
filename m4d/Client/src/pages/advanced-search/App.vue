<template>
  <page
    id="app"
    :consumesEnvironment="true"
    :consumesTags="true"
    @tag-database-loaded="onEnvironmentLoaded"
  >
    <h1 class="col-sm-12" style="font-size: 22px; text-align: center">
      Advanced Song Search
    </h1>
    <div style="max-width: 600px; margin-left: auto; margin-right: auto">
      <b-form
        id="advanced-search"
        @submit.stop.prevent="onSubmit"
        @reset="onReset"
        :validated="validated"
        novalidate
      >
        <b-form-group
          id="searchStringGroup"
          label="Keywords:"
          label-for="searchString"
        >
          <b-form-input
            id="searchString"
            v-model="keyWords"
            type="text"
            placeholder="Enter Keywords from title, artist, etc... OR a Spotify or Apple share link"
            ref="keywords"
            @input="checkServiceAndWarn"
          ></b-form-input>
        </b-form-group>

        <b-form-group label="Dances:">
          <div style="border: 1px solid #ced4da; boder-radius: 0.25rem">
            <dance-selector
              :danceList="model.dances"
              v-model="dances"
            ></dance-selector>
            <div class="d-flex justify-content-between w-100 mx-1 mb-2">
              <b-form-radio-group
                id="dance-connector"
                v-model="danceConnector"
                buttons
                button-variant="outline-secondary"
                :disabled="dances.length < 2"
                class="mx-3"
                size="sm"
              >
                <b-form-radio value="any">Any</b-form-radio>
                <b-form-radio value="all">All</b-form-radio>
              </b-form-radio-group>
            </div>
          </div>
        </b-form-group>

        <b-form-group label="Include Tags:">
          <tag-category-selector
            id="includeTags"
            :tagList="tags"
            chooseLabel="Choose Tags to Include"
            searchLabel="Search Tags"
            emptyLabel="No more tags to choose"
            v-model="includeTags"
          ></tag-category-selector>
        </b-form-group>

        <b-form-group label="Exclude Tags:">
          <tag-category-selector
            id="excludeTags"
            :tagList="tags"
            chooseLabel="Choose Tags to Exclude"
            searchLabel="Search Tags"
            emptyLabel="No more tags to choose"
            v-model="excludeTags"
          ></tag-category-selector>
        </b-form-group>

        <b-form-group
          id="tempo-range-group"
          label="Tempo range (BPM):"
          label-for="tempo-range"
        >
          <b-form-group id="tempo-range">
            <div class="d-flex">
              <b-form-input
                id="tempo-min"
                v-model="tempoMin"
                type="number"
                number
                min="0"
                max="250"
                style="width: 6rem"
              ></b-form-input>
              <span class="mx-2 pt-2">to</span>
              <b-form-input
                id="tempo-max"
                v-model="tempoMax"
                type="number"
                min="0"
                max="250"
                number
                style="width: 6rem"
              ></b-form-input>
              <div class="invalid-feedback">
                Tempos must be between 0 and 250 BPM
              </div>
            </div>
          </b-form-group>
        </b-form-group>

        <b-form-group id="activity-group" label="By User:" label-for="activity">
          <div class="d-flex">
            <b-form-input
              id="user"
              placeholder="UserName"
              v-model="displayUser"
              style="width: 10rem"
              class="mr-2"
            ></b-form-input>
            <b-form-select
              id="activity"
              v-model="computedActivity"
              :options="activities"
            ></b-form-select>
          </div>
        </b-form-group>

        <b-form-group
          id="services-group"
          class="mx-2 mb-2"
          label="Available on:"
          label-for="services"
        >
          <b-form-checkbox-group v-model="services" id="services">
            <b-form-checkbox value="A">Amazon</b-form-checkbox>
            <b-form-checkbox value="I">ITunes</b-form-checkbox>
            <b-form-checkbox value="S">Spotify</b-form-checkbox>
          </b-form-checkbox-group>
        </b-form-group>

        <b-form-group
          id="bonuses-group"
          class="mx-2 mb-2"
          label="Bonus content:"
          label-for="bonuses"
        >
          <b-form-checkbox-group v-model="bonuses" id="bonuses">
            <b-form-checkbox value="P"
              >Not found in any publisher catalog</b-form-checkbox
            >
            <b-form-checkbox value="D"
              >Not categorized by dance</b-form-checkbox
            >
          </b-form-checkbox-group>
          <template v-slot:description>
            <a href="https://music4dance.blog/music4dance-help/subscriptions/"
              >Premium content</a
            >
          </template>
        </b-form-group>

        <b-form-group id="sort-group" label="Sort By:" label-for="sort">
          <b-form-select
            id="sort"
            v-model="sort"
            :options="sortOptions"
            required
          ></b-form-select>
          <b-form-radio-group
            id="sort-order"
            v-model="order"
            name="sort-order"
            class="mt-2"
          >
            <b-form-radio value="asc"
              >Ascending (A-Z, Slow-Fast, Newest-Oldest)</b-form-radio
            >
            <b-form-radio value="desc"
              >Descending (Z-A, Fast-Slow, Oldest-Newest)</b-form-radio
            >
          </b-form-radio-group>
        </b-form-group>

        <div class="d-flex justify-content-between w-100 mx-1 mb-2">
          <b-button type="reset" variant="secondary">Reset</b-button>
          <b-button type="search" variant="primary">Submit</b-button>
        </div>
      </b-form>
      <b-card class="mt-3" header="Form Data Result" v-if="showDiagnostics">
        <pre class="m-0">
searchString = {{ keyWords }}
dances = {{ dances }}
danceConnector = {{ danceConnector }}
tempoMin = {{ tempoMin }}
tempoMax = {{ tempoMax }}
activity = {{ activity }}
services = {{ services }}
sort = {{ sort }}
order = {{ order }}
bonus = {{ bonuses }}
includeTags = {{ includeTags }}
excludeTags = {{ excludeTags }}

filter = {{ songFilter }}
      </pre
        >
      </b-card>
    </div>
  </page>
</template>

<script lang="ts">
import DanceSelector from "@/components/DanceSelector.vue";
import Page from "@/components/Page.vue";
import TagCategorySelector from "@/components/TagCategorySelector.vue";
import AdminTools from "@/mix-ins/AdminTools";
import DropTarget from "@/mix-ins/DropTarget";
import EnvironmentManager from "@/mix-ins/EnvironmentManager";
import { DanceQuery } from "@/model/DanceQuery";
import { SongFilter } from "@/model/SongFilter";
import { SongSort, SortOrder } from "@/model/SongSort";
import { Tag } from "@/model/Tag";
import { TagDatabase } from "@/model/TagDatabase";
import { UserQuery } from "@/model/UserQuery";
import "reflect-metadata";
import { TypedJSON } from "typedjson";
import { Component, Mixins, Vue } from "vue-property-decorator";
import { SearchModel } from "./searchModel";

declare const model: SearchModel;

@Component({
  components: {
    DanceSelector,
    Page,
    TagCategorySelector,
  },
})
export default class App extends Mixins(
  AdminTools,
  EnvironmentManager,
  DropTarget
) {
  private showDiagnostics = false;
  private keyWords = "";

  private dances: string[] = [];
  private danceConnector = "any";
  private tags: Tag[] = [];

  private includeTags: string[] = [];
  private excludeTags: string[] = [];

  private tempoMin = 0;
  private tempoMax = 250;

  private user: string;
  private displayUser: string;

  private activity = "NT";
  private get computedActivity(): string {
    return this.displayUser ? this.activity : "NT";
  }
  private set computedActivity(value: string) {
    if (this.hasUser) {
      this.activity = value;
    }
  }

  private get hasUser(): boolean {
    return !!(this.user || this.displayUser);
  }

  private get isAnonymous(): boolean {
    return this.displayUser === "Anonymous";
  }

  private get activities() {
    const user = this.displayUser;
    const empty = { text: "Don't filter on user activity", value: "NT" };
    const my = user === "me" ? "my" : user + "'s";
    const i = user === "me" ? "I have" : user + " has";

    return user
      ? [
          empty,
          { text: `Include all songs in ${my} favorites`, value: "IL" },
          { text: `Exclude all songs in ${my} favorites`, value: "XL" },
          { text: `Include all songs ${i}  tagged`, value: "IT" },
          { text: `Exclude all songs ${i} tagged`, value: "XT" },
          { text: `Exclude all songs in ${my} blocked list`, value: "XH" },
          { text: `Include all songs in ${my} blocked list`, value: "IH" },
        ]
      : [empty];
  }

  private services: string[] = [];

  private sortOptions = [
    { text: "Title", value: SortOrder.Title },
    { text: "Artist", value: SortOrder.Artist },
    { text: "Tempo", value: SortOrder.Tempo },
    { text: "Dance Rating", value: SortOrder.Dances },
    { text: "Last Modified", value: SortOrder.Modified },
    { text: "Last Edited", value: SortOrder.Edited },
    { text: "When Added", value: SortOrder.Created },
    { text: "Energy", value: SortOrder.Energy },
    { text: "Mood", value: SortOrder.Mood },
    { text: "Strength of Beat", value: SortOrder.Beat },
    { text: "Closest Match", value: null },
  ];
  private sort: string | null = "Dances";
  private order = "asc";
  private bonuses: string[] = [];
  private model: SearchModel;
  private validated = false;

  constructor() {
    super();

    this.model = TypedJSON.parse(model, SearchModel)!;
    const filter = this.getSourceFilter(model);
    const danceQuery = new DanceQuery(filter.dances);

    this.keyWords = filter.searchString ?? "";

    this.dances = danceQuery.danceList;
    this.danceConnector = danceQuery.isExclusive ? "all" : "any";

    const sort = new SongSort(filter.sortOrder);
    this.sort = sort.order ?? null;
    this.order = sort.direction;

    this.services = filter.purchase ? filter.purchase.trim().split("") : [];

    // Real initialization of user defered to after environment is loaded
    this.activity = "NT";
    this.user = "";
    this.displayUser = "";

    this.tempoMin = filter.tempoMin ?? 0;
    this.tempoMax = filter.tempoMax ?? 250;

    this.includeTags = filter.tags ? this.extractTags(filter.tags, true) : [];
    this.excludeTags = filter.tags ? this.extractTags(filter.tags, false) : [];

    this.bonuses = [];
    if (filter.level && filter.level & 1) {
      this.bonuses.push("P");
    }
    if (filter.level && filter.level & 2) {
      this.bonuses.push("D");
    }
  }

  private get tempoValid(): boolean {
    return this.tempoMin <= this.tempoMax;
  }

  private get songFilter(): SongFilter {
    const danceQuery = DanceQuery.fromParts(
      this.dances,
      this.danceConnector === "all"
    );
    const userQuery = UserQuery.fromParts(
      this.computedActivity ? this.computedActivity : undefined,
      this.isAnonymous ? this.user : this.displayUser
    );
    const filter = new SongFilter();
    let level = 0;
    if (this.bonuses.indexOf("P") !== -1) {
      level = 1;
    }
    if (this.bonuses.indexOf("D") !== -1) {
      level += 2;
    }

    filter.action = "Advanced";
    filter.searchString = this.keyWords;
    filter.dances = danceQuery.query;
    filter.sortOrder = SongSort.fromParts(
      this.sort ?? undefined,
      this.order
    ).query;
    filter.user = userQuery.query;
    filter.purchase = this.services.join("");
    filter.tempoMin = this.tempoMin === 0 ? undefined : this.tempoMin;
    filter.tempoMax = this.tempoMax >= 250 ? undefined : this.tempoMax;
    filter.tags = this.buildTagList();
    filter.level = level ? level : undefined;

    return filter;
  }

  private getQueryFilter(): SongFilter | undefined {
    const params = new URLSearchParams(window.location.search);
    const filterString = params.get("filter");

    return filterString ? SongFilter.buildFilter(filterString) : undefined;
  }

  private getSourceFilter(model?: SearchModel): SongFilter {
    const queryFilter = this.getQueryFilter();
    model = model ? model : this.model;
    return queryFilter ? queryFilter : model.filter;
  }

  private async onSubmit(): Promise<void> {
    const form = document.getElementById("advanced-search") as HTMLFormElement;

    if (form.checkValidity() === true) {
      if (this.tempoMin > this.tempoMax) {
        const tempo = this.tempoMax;
        this.tempoMax = this.tempoMin;
        this.tempoMin = tempo;
      }

      const loc = window.location;
      const query = this.songFilter.encodedQuery;

      window.location.href = `${loc.origin}/song/filtersearch?filter=${query}`;
    }

    this.validated = true;
  }

  private onReset(evt: Event) {
    evt.preventDefault();
    const userName = this.isAnonymous ? this.userName : this.displayUser;

    this.keyWords = "";
    this.dances.splice(0);
    this.danceConnector = "any";
    this.includeTags.splice(0);
    this.excludeTags.splice(0);
    this.tempoMin = 0;
    this.tempoMax = 250;
    if (userName) {
      this.user = userName;
      this.activity = "IH";
    } else {
      this.user = "";
      this.activity = "NT";
    }
    this.displayUser = "";
    this.services.splice(0);
    this.sort = "Dances";
    this.order = "asc";
    this.bonuses.splice(0);

    this.validated = false;
  }

  private buildTagList(): string {
    const lists: string[] = [];
    if (this.includeTags.length > 0) {
      lists.push(this.buildSingleTagList(this.includeTags, "+"));
    }
    if (this.excludeTags.length > 0) {
      lists.push(this.buildSingleTagList(this.excludeTags, "-"));
    }
    return lists.join("|");
  }

  private buildSingleTagList(tags: string[], decorator: string) {
    return tags.map((t) => `${decorator}${t}`).join("|");
  }

  private extractTags(tags: string, include: boolean): string[] {
    if (!tags) {
      return [];
    }

    const qualifier = include ? "+" : "-";
    const parts = tags.split("|").map((p) => p.trim());
    let filtered = parts
      .filter((p) => p.startsWith(qualifier))
      .map((p) => p.slice(1));
    if (include) {
      filtered = filtered.concat(
        parts.filter((p) => !p.startsWith("+") && !p.startsWith("-"))
      );
    }

    return filtered;
  }

  private async onEnvironmentLoaded(tagDatabase: TagDatabase): Promise<void> {
    const filter = this.getSourceFilter();
    const userQuery = new UserQuery(filter.user);
    this.activity = userQuery.parts;
    this.user = userQuery.userName ?? this.userName ?? "";
    this.displayUser = userQuery.displayName;
    this.tags = tagDatabase.tags;

    await this.$nextTick();
    ((this.$refs.keywords as Vue).$el as HTMLElement).focus();
  }
}
</script>
