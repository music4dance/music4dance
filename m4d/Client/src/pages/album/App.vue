<template>
  <page id="app">
    <h1>
      Album: {{ model.title }}
      <span v-if="model.artist"
        >by <a :href="artistRef">{{ model.artist }}</a></span
      >
    </h1>
    <song-table
      :histories="model.histories"
      :filter="model.filter"
      :hideSort="true"
      :hiddenColumns="hidden"
      @song-selected="selectSong"
    ></song-table>
    <admin-footer :model="model" :selected="selected"></admin-footer>
  </page>
</template>

<script lang="ts">
import Page from "@/components/Page.vue";
import SongTable from "@/components/SongTable.vue";
import AdminFooter from "@/components/AdminFooter.vue";
import { safeEnvironment } from "@/helpers/DanceEnvironmentManager";
import { AlbumModel } from "@/model/AlbumModel";
import { DanceEnvironment } from "@/model/DanceEnvironment";
import "reflect-metadata";
import { TypedJSON } from "typedjson";
import SongSelector from "@/mix-ins/SongSelector";

declare const model: string;

export default SongSelector.extend({
  components: { AdminFooter, Page, SongTable },
  data() {
    return new (class {
      model: AlbumModel = TypedJSON.parse(model, AlbumModel)!;
      environment: DanceEnvironment = safeEnvironment();
    })();
  },
  computed: {
    artistRef(): string {
      return `/song/artist?name=${encodeURIComponent(this.model.artist || "")}`;
    },
    hidden(): string[] {
      return this.model.artist ? ["Artist"] : [];
    },
  },
});
</script>
