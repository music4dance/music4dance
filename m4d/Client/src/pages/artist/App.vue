<template>
  <page id="app" :title="title" :consumesEnvironment="true">
    <song-table 
      :songs="model.songs"
      :filter="model.filter"
      :userName="model.userName"
      :hideSort="true"
      :hiddenColumns="['Artist']"
    ></song-table>
  </page>
</template>

<script lang="ts">
// tslint:disable: max-classes-per-file
import 'reflect-metadata';
import { Component, Vue } from 'vue-property-decorator';
import SongTable from '@/components/SongTable.vue';
import Page from '@/components/Page.vue';
import { TypedJSON } from 'typedjson';
import { Song } from '@/model/Song';
import { ArtistModel } from '@/model/ArtistModel';

declare const model: string;

@Component({
  components: {
    Page,
    SongTable,
  },
})
export default class App extends Vue {
  private readonly model: ArtistModel;

  constructor() {
    super();

    this.model = TypedJSON.parse(model, ArtistModel)!;
  }

  private get title(): string {
    return `Artist: ${this.model.artist}`;
  }
}
</script>
