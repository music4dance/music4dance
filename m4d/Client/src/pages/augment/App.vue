<template>
  <page
    id="app"
    :consumesEnvironment="true"
    @environment-loaded="onEnvironmentLoaded"
  >
    <b-row v-if="phase === 'lookup'">
      <b-col>
        <b-alert v-if="lastSong" dismissible show="10">
          Thank you for {{ created ? "adding" : "editing" }}
          <i>{{ lastSong.title }}</i> by {{ lastSong.artist }}
        </b-alert>
        <h1>Add Song</h1>
        <b-tabs v-if="canAugment" v-model="tabIndex" card>
          <b-tab title="by Title"
            ><b-card-text
              ><augment-search
                @edit-song="editSong($event)"
              ></augment-search></b-card-text
          ></b-tab>
          <b-tab title="by Id">
            <b-card-text
              ><augment-lookup @edit-song="editSong($event)"></augment-lookup
            ></b-card-text>
          </b-tab>
          <b-tab v-if="isAdmin" title="Admin">
            <b-card-text>
              <p>Paste in text version of propery list</p>
              <b-input-group prepend="Properties">
                <b-form-input
                  id="admin-properties"
                  palceholder="Song Properties TSV"
                  v-model="propertiesString"
                  trim
                ></b-form-input>
                <b-button variant="primary" @click="adminCreate"
                  >Create</b-button
                >
              </b-input-group>
            </b-card-text>
          </b-tab>
        </b-tabs>
        <augment-info v-else> </augment-info>
      </b-col>
    </b-row>
    <div v-else>
      <b-alert show variant="success">
        <div v-if="songModel.created">
          <b>Create Song:</b> This song is new to music4dance, please fill in
          missing fields and click <b>Add Song</b> to add this song to the
          catalog. Remember, you must vote on at least one dance to add this
          song.
        </div>
        <div v-else>
          <b>Edit Song:</b> We found this song in the music4dance catalog,
          please vote on dance styles and add tags to improve the catalog.
        </div>
      </b-alert>
      <song-core
        :model="songModel"
        :environment="environment"
        :startEditing="true"
        :creating="songModel.created"
        @song-saved="reset(true)"
        @cancel-changes="reset(false)"
      ></song-core>
    </div>
  </page>
</template>

<script lang="ts">
import "reflect-metadata";
import { Component, Mixins } from "vue-property-decorator";
import AdminTools from "@/mix-ins/AdminTools";
import AugmentInfo from "./components/AugmentInfo.vue";
import AugmentLookup from "./components/AugmentLookup.vue";
import AugmentSearch from "./components/AugmentSearch.vue";
import SongCore from "@/pages/song/components/SongCore.vue";
import Page from "@/components/Page.vue";
import { SongDetailsModel } from "@/model/SongDetailsModel";
import { SongFilter } from "@/model/SongFilter";
import { DanceEnvironment } from "@/model/DanceEnvironmet";
import { Song } from "@/model/Song";
import { SongHistory } from "@/model/SongHistory";

enum AugmentPhase {
  lookup = "lookup",
  results = "results",
  edit = "edit",
}

@Component({
  components: { AugmentInfo, AugmentLookup, AugmentSearch, SongCore, Page },
})
export default class App extends Mixins(AdminTools) {
  private phase = AugmentPhase.lookup;
  private songModel: SongDetailsModel | null = null;
  private lastSong: Song | null = null;
  private created = false;
  private environment: DanceEnvironment | null = null;
  private propertiesString = "";
  private tabIndex = 0;

  private get canAugment(): boolean {
    return this.canTag || this.isPremium;
  }

  private editSong(model: SongDetailsModel): void {
    this.songModel = new SongDetailsModel({
      created: !!model.created,
      songHistory: model.songHistory,
      filter: new SongFilter(),
      userName: this.userName,
    });
    this.phase = AugmentPhase.edit;
  }

  private async adminCreate(): Promise<void> {
    console.log(`Create Song: ${this.propertiesString}`);
    this.songModel = new SongDetailsModel({
      created: true,
      songHistory: SongHistory.fromString(this.propertiesString),
      filter: new SongFilter(),
      userName: this.userName,
    });
    this.propertiesString = "";
    this.phase = AugmentPhase.edit;
  }

  private reset(saved: boolean): void {
    if (saved) {
      this.lastSong = Song.fromHistory(this.songModel!.songHistory);
      this.created = !!this.songModel!.created;
    }
    this.phase = AugmentPhase.lookup;
    this.songModel = null;
  }

  private onEnvironmentLoaded(environment: DanceEnvironment): void {
    this.environment = environment;
  }
}
</script>

<style lang="scss" scoped>
#service-id {
  width: 25rem;
}
</style>
