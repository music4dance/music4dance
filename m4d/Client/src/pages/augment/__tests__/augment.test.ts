import { Preloads } from "@/Preloads";
import { testPageAsync } from "@/helpers/TestHelpers";
import App from "../App.vue";

describe("Augment.vue", () => {
  test("renders the augment page", async () => {
    await testPageAsync(App, Preloads.All, {});
  });
});
