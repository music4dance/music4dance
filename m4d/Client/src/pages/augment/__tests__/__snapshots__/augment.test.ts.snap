// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Augment.vue renders the augment page 1`] = `
<div>
  <div>
    <!---->
    <!---->
    <nav class="navbar navbar-dark bg-primary navbar-expand-lg" id="mainMenu"><a href="/" target="_self" class="navbar-brand"><img src="/images/header-logo.png" height="40" title="music4dance"></a> <button type="button" aria-label="Toggle navigation" class="navbar-toggler collapsed" aria-expanded="false" aria-controls="nav-collapse"><span class="navbar-toggler-icon"></span></button>
      <transition-stub css="true" enterclass="" leaveclass="collapse show" entertoclass="collapse show" leavetoclass="collapse" enteractiveclass="collapsing" leaveactiveclass="collapsing">
        <div id="nav-collapse" class="navbar-collapse collapse" style="display: none;">
          <ul class="navbar-nav">
            <li id="music-menu" class="nav-item b-nav-dropdown dropdown"><a id="music-menu__BV_toggle_" role="button" aria-haspopup="true" aria-expanded="false" aria-controls="music-menu__BV_toggle_menu_" href="#music-menu" target="_self" class="nav-link dropdown-toggle"><span>Music</span></a>
              <ul tabindex="-1" aria-labelledby="music-menu__BV_toggle_" id="music-menu__BV_toggle_menu_" class="dropdown-menu">
                <li role="presentation"><a role="menuitem" href="/dances" target="_self" class="dropdown-item">Dances</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/ballroom-competition-categories" target="_self" class="dropdown-item">Ballroom</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/latin" target="_self" class="dropdown-item">Latin</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/swing" target="_self" class="dropdown-item">Swing</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/tango" target="_self" class="dropdown-item">Tango</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/dances/country" target="_self" class="dropdown-item">Country</a></li>
                <li role="presentation"><a role="menuitem" href="/song" target="_self" class="dropdown-item">Song Library</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/song/advancedsearchform" target="_self" class="dropdown-item">Advanced Search</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/song/augment" target="_self" class="dropdown-item">Add Song</a></li>
                <li role="presentation" class="nav-subitem"><a role="menuitem" href="/song/newmusic" target="_self" class="dropdown-item">New Music</a></li>
                <li role="presentation"><a role="menuitem" href="/dances/wedding-music" target="_self" class="dropdown-item">Wedding</a></li>
                <li role="presentation"><a role="menuitem" href="/song/holidaymusic" target="_self" class="dropdown-item">Holiday</a></li>
                <li role="presentation"><a role="menuitem" href="/song/holidaymusic?occassion=halloween" target="_self" class="dropdown-item">Halloween</a></li>
                <li role="presentation"><a role="menuitem" href="/tag" target="_self" class="dropdown-item">Tags</a></li>
              </ul>
            </li>
            <li id="tools-menu" class="nav-item b-nav-dropdown dropdown"><a id="tools-menu__BV_toggle_" role="button" aria-haspopup="true" aria-expanded="false" aria-controls="tools-menu__BV_toggle_menu_" href="#tools-menu" target="_self" class="nav-link dropdown-toggle"><span>Tools</span></a>
              <ul tabindex="-1" aria-labelledby="tools-menu__BV_toggle_" id="tools-menu__BV_toggle_menu_" class="dropdown-menu">
                <li role="presentation"><a role="menuitem" href="/home/counter" target="_self" class="dropdown-item">Tempo Counter</a></li>
                <li role="presentation"><a role="menuitem" href="/home/tempi" target="_self" class="dropdown-item">Tempi (Tempos)</a></li>
                <li role="presentation"><a role="menuitem" href="/song/advancedsearchform" target="_self" class="dropdown-item">Advanced Search</a></li>
                <!---->
              </ul>
            </li>
            <li class="nav-item b-nav-dropdown dropdown"><a role="button" aria-haspopup="true" aria-expanded="false" href="#" target="_self" class="nav-link dropdown-toggle"><span>Info</span></a>
              <ul tabindex="-1" class="dropdown-menu">
                <li role="presentation"><a role="menuitem" href="https://music4dance.blog/music4dance-help/song-list/" target="_self" class="dropdown-item">Help</a></li>
                <li role="presentation"><a role="menuitem" href="https://music4dance.blog/" target="_self" class="dropdown-item">Blog</a></li>
                <li role="presentation"><a role="menuitem" href="/home/faq" target="_self" class="dropdown-item">FAQ</a></li>
                <li role="presentation"><a role="menuitem" href="/home/about" target="_self" class="dropdown-item">About music4dance</a></li>
                <li role="presentation"><a role="menuitem" href="/home/readinglist" target="_self" class="dropdown-item">Reading List</a></li>
                <li role="presentation"><a role="menuitem" href="/home/sitemap" target="_self" class="dropdown-item">Site Map</a></li>
                <li role="presentation"><a role="menuitem" href="/home/privacypolicy" target="_self" class="dropdown-item">Privacy Policy</a></li>
                <li role="presentation"><a role="menuitem" href="/home/credits" target="_self" class="dropdown-item">Credits</a></li>
              </ul>
            </li>
            <li id="contribute-menu" right="" class="nav-item"><a href="/home/contribute" target="_self" class="nav-link">Contribute</a></li>
            <!---->
          </ul>
          <ul class="navbar-nav ml-auto">
            <li class="form-inline">
              <form class="form-inline"><input type="text" placeholder="Search" class="mr-sm-2 form-control form-control-sm"> <button type="submit" disabled="disabled" class="btn my-2 my-sm-0 btn-secondary btn-sm disabled">Search</button></form>
            </li>
            <li right="" class="nav-item"><a href="/identity/account/register?returnUrl=/" target="_self" class="nav-link">Register</a></li>
            <li right="" class="nav-item"><a href="/identity/account/login?returnUrl=/" target="_self" class="nav-link">Login</a></li>
          </ul>
        </div>
      </transition-stub>
    </nav>
    <!---->
    <!---->
    <form id="logoutForm" action="/identity/account/logout" method="post" style="height: 0px;"><input name="__RequestVerificationToken" type="hidden" value="FOO"> <input type="hidden" name="returnUrl" value="/"> <button id="logout" type="submit" class="btn btn-link"></button></form>
  </div>
  <!---->
  <div id="body-content" class="container-fluid body-content">
    <div class="mt-2"></div>
    <div>
      <div class="row">
        <div class="col">
          <!---->
          <h1>Add Song</h1>
          <div>
            <p>
              In order to add songs, you must first
              <a href="/identity/account/register?returnUrl=/song/augment">register</a>
              and
              <a href="/identity/account/login?returnUrl=/song/augment">sign in.</a>
            </p>
          </div>
        </div>
      </div>
      <div class="row">
        <div class="col">
          <transition-stub name="" css="true" enterclass="" leaveclass="show" entertoclass="show" leavetoclass="" enteractiveclass="" leaveactiveclass="">
            <div role="alert" aria-live="polite" aria-atomic="true" class="alert alert-warning">
              <!---->
              Adding songs is a new feature so if you run into bugs or have
              suggestions for improving this feature please don't hesitate to send
              email to
              <a href="mailto:info@music4dance.net">info@music4dance.net</a> or fill
              out our
              <a href="https://music4dance.blog/feedback/">feedback form</a> and
              we'll be happy to take your input as we contineu to improve the
              feature.
            </div>
          </transition-stub>
        </div>
      </div>
    </div>
  </div>
  <div id="footer-content">
    <hr>
    <footer>
      <p>
        Â© 2022 -
        <a href="https://www.music4dance.net">Music4Dance.net</a> -
        <a href="https://www.music4dance.net/home/sitemap">Site Map</a> -
        <a href="https://www.music4dance.net/home/termsofservice">Terms of Service</a>
        -
        <a href="https://www.music4dance.net/home/privacypolicy">Privacy Policy</a>
        - <a href="https://www.music4dance.net/home/credits">Credits</a> -
        <a href="https://music4dance.blog/music4dance-help/song-list/">Help</a>
      </p>
    </footer>
  </div>
  <!---->
</div>
`;
